### ISSUE TITLE
Error creating ALSA sequencer client object.

### ISSUE THREAD

**Issue by OPNA2608** [rerrahkr/BambooTracker#172]
On my main system (Linux, NixOS 19.09), after a fresh boot, I'm unable to start BambooTracker. It aborts with a SIGSEGV and prints this:
```
Translation "qt_de_DE" from "/nix/store/p16qsg4ram58g9aalkd0dixf000jgzbh-qtbase-5.12.3/translations"
Translation "bamboo_tracker_de_DE" from "/home/bt1cn/devel/BambooTracker/BambooTracker/BambooTracker/.qm"
Using emulator: Nuked OPN-Mod
ALSA lib seq_hw.c:466:(snd_seq_hw_open) open /dev/snd/seq failed: Permission denied

MidiInAlsa::initialize: error creating ALSA sequencer client object.

Segmentation fault (Speicherabzug geschrieben)
```
(last line means process aborted with a segmentation fault)

If I start the app with `sudo` once and try again, it works fine. Same if I add myself to the `audio` group and log out and back in.
It's possible my system is misconfigured and I would indeed need the `audio` group or proper permissions for /dev/snd/seq, however I have configured the use of the Pulse backend instead so the error is a bit irritation. I guess this comes from the MIDI initialization? Furthermore, on my Ubuntu machine I'm not in the `audio` group either yet it starts up just fine there.

Here's the debugger report on the actual SIGSEGV:
```
MidiInAlsa::initialize: error creating ALSA sequencer client object.

==6731== Invalid read of size 8
==6731==    at 0x5257E08: ??? (in /nix/store/p16qsg4ram58g9aalkd0dixf000jgzbh-qtbase-5.12.3/lib/libQt5Gui.so.5.12.3)
==6731==    by 0x5259204: QGuiApplication::font() (in /nix/store/p16qsg4ram58g9aalkd0dixf000jgzbh-qtbase-5.12.3/lib/libQt5Gui.so.5.12.3)
==6731==    by 0x5311770: QFont::QFont() (in /nix/store/p16qsg4ram58g9aalkd0dixf000jgzbh-qtbase-5.12.3/lib/libQt5Gui.so.5.12.3)
==6731==    by 0x4C5B580: QWidgetPrivate::QWidgetPrivate(int) (in /nix/store/p16qsg4ram58g9aalkd0dixf000jgzbh-qtbase-5.12.3/lib/libQt5Widgets.so.5.12.3)
==6731==    by 0x4E56B68: QMessageBox::QMessageBox(QMessageBox::Icon, QString const&, QString const&, QFlags<QMessageBox::StandardButton>, QWidget*, QFlags<Qt::WindowType>) (in /nix/store/p16qsg4ram58g9aalkd0dixf000jgzbh-qtbase-5.12.3/lib/libQt5Widgets.so.5.12.3)
==6731==    by 0x4E56F0D: ??? (in /nix/store/p16qsg4ram58g9aalkd0dixf000jgzbh-qtbase-5.12.3/lib/libQt5Widgets.so.5.12.3)
==6731==    by 0x4264D6: main.cold.63 (main.cpp:38)
==6731==    by 0x63EEB8D: (below main) (in /nix/store/pnd2kl27sag76h23wa5kl95a76n3k9i3-glibc-2.27/lib/libc-2.27.so)
==6731==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
==6731== 
==6731== 
==6731== Process terminating with default action of signal 11 (SIGSEGV): dumping core
==6731==  Access not within mapped region at address 0x0
==6731==    at 0x5257E08: ??? (in /nix/store/p16qsg4ram58g9aalkd0dixf000jgzbh-qtbase-5.12.3/lib/libQt5Gui.so.5.12.3)
==6731==    by 0x5259204: QGuiApplication::font() (in /nix/store/p16qsg4ram58g9aalkd0dixf000jgzbh-qtbase-5.12.3/lib/libQt5Gui.so.5.12.3)
==6731==    by 0x5311770: QFont::QFont() (in /nix/store/p16qsg4ram58g9aalkd0dixf000jgzbh-qtbase-5.12.3/lib/libQt5Gui.so.5.12.3)
==6731==    by 0x4C5B580: QWidgetPrivate::QWidgetPrivate(int) (in /nix/store/p16qsg4ram58g9aalkd0dixf000jgzbh-qtbase-5.12.3/lib/libQt5Widgets.so.5.12.3)
==6731==    by 0x4E56B68: QMessageBox::QMessageBox(QMessageBox::Icon, QString const&, QString const&, QFlags<QMessageBox::StandardButton>, QWidget*, QFlags<Qt::WindowType>) (in /nix/store/p16qsg4ram58g9aalkd0dixf000jgzbh-qtbase-5.12.3/lib/libQt5Widgets.so.5.12.3)
==6731==    by 0x4E56F0D: ??? (in /nix/store/p16qsg4ram58g9aalkd0dixf000jgzbh-qtbase-5.12.3/lib/libQt5Widgets.so.5.12.3)
==6731==    by 0x4264D6: main.cold.63 (main.cpp:38)
==6731==    by 0x63EEB8D: (below main) (in /nix/store/pnd2kl27sag76h23wa5kl95a76n3k9i3-glibc-2.27/lib/libc-2.27.so)
```

**Comment by jpcima** [comment]
>I guess this comes from the MIDI initialization?

Hi. The seq interface is MIDI and strictly unrelated to Pulseaudio.

Did you check what are permissions of this device node, with such a command as `ls -l /dev/snd/seq`? It will be `root:audio` and `rw-rw----` usually.

**Comment by OPNA2608** [comment]
```
crw-rw---- 1 root audio 116, 1 16. Dez 19:29 /dev/snd/seq
```

**Comment by jpcima** [comment]
In this case, this error code has to be emitted from the driver or some part of the Linux kernel.
Without a doubt, it's not a fault of Bamboo Tracker.

You could check if you have a sort of Linux security module running, which is at the origin of the access problem, like Selinux or another. Or, try if it's working on another kernel.

**Comment by OPNA2608** [comment]
> You could check if you have a sort of Linux security module running, which is at the origin of the access problem, like Selinux or another.

Checked for either AppArmor or SELinux, nada.
```
root@NixTop:~# aa-status
The program ‘aa-status’ is currently not installed. You can install it by typing:
  nix-env -iA nixos.apparmor-utils

root@NixTop:~# nix-env -iA nixos.apparmor-utils
installing 'apparmor-utils-2.13.1'

root@NixTop:~# aa-status
apparmor module is loaded.
0 profiles are loaded.
0 profiles are in enforce mode.
0 profiles are in complain mode.
0 processes have profiles defined.
0 processes are in enforce mode.
0 processes are in complain mode.
0 processes are unconfined but have a profile defined.

root@NixTop:~# sestatus
The program ‘sestatus’ is currently not installed. You can install it by typing:
  nix-env -iA nixos.policycoreutils

root@NixTop:~# nix-env -iA nixos.policycoreutils
installing 'policycoreutils-2.9'

root@NixTop:~# sestatus
SELinux status:                 disabled
```

Ubuntu machine is running AppArmor, but it basically only lists `man` and `libreoffice` stuff there.

> Or, try if it's working on another kernel.

Running kernel `4.14.157`, Ubuntu machine `4.15.0-55-generic`. I can't mess with the kernel version too much without reworking my system configuration quite abit, limited by novideo driver compatibility and the selection of kernel versions on the stable release of my distro.

---

I have pinned it down to [this very line](https://github.com/rerrahkr/BambooTracker/blob/8e5357eddb769b8a07b0fce1dfde24057f2c6488/BambooTracker/midi/midi.cpp#L36). I'm aware that the seq interface is strictly for MIDI and unrelated to the Pulse selection in BT. I was trying to say that it occurs while initializing the MIDI-In and MIDI-Out, which still goes via ALSA when using Pulse.
```cpp
RtMidiIn *inputClient = new RtMidiIn(api, MIDI_INP_CLIENT_NAME, MidiBufferSize);
```

[This looks related?](https://github.com/captnfab/PianoBooster/issues/150) The solution was to wrap the `RtMidiIn()` and `RtMidiOut()` and handle the error. If I do something similar, I *don't* get the SIGSEGV anymore:
```patch
diff --git a/BambooTracker/midi/midi.cpp b/BambooTracker/midi/midi.cpp
index 1eee0d3..fa17f72 100644
--- a/BambooTracker/midi/midi.cpp
+++ b/BambooTracker/midi/midi.cpp
@@ -1,6 +1,7 @@
 #include "midi.hpp"
 #include "midi_def.h"
 #include <stdio.h>
+#include <iostream>
 
 std::unique_ptr<MidiInterface> MidiInterface::instance_;
 
@@ -33,7 +34,13 @@ void MidiInterface::switchApi(RtMidi::Api api)
        if (inputClient_ && api == inputClient_->getCurrentApi())
                return;
 
-       RtMidiIn *inputClient = new RtMidiIn(api, MIDI_INP_CLIENT_NAME, MidiBufferSize);
+  RtMidiIn *inputClient;
+  try {
+         inputClient = new RtMidiIn(api, MIDI_INP_CLIENT_NAME, MidiBufferSize);
+  } catch (RtMidiError &error) {
+    std::cout << "DGHJGKJHGNSDFGH#################" << std::endl;
+    exit (1);
+  }
        inputClient->ignoreTypes(MIDI_INP_IGNORE_SYSEX, MIDI_INP_IGNORE_TIME, MIDI_INP_IGNORE_SENSE);
        inputClient_.reset(inputClient);
        inputClient->setErrorCallback(&onMidiError, this);
```
```
[nix-shell:~/devel/BambooTracker/BambooTracker/BambooTracker]$ ./BambooTracker 
Translation "qt_de_DE" from "/nix/store/p16qsg4ram58g9aalkd0dixf000jgzbh-qtbase-5.12.3/translations"
Translation "bamboo_tracker_de_DE" from "/home/bt1cn/devel/BambooTracker/BambooTracker/BambooTracker/.qm"
Using emulator: Nuked OPN-Mod
ALSA lib seq_hw.c:466:(snd_seq_hw_open) open /dev/snd/seq failed: Permission denied

MidiInAlsa::initialize: error creating ALSA sequencer client object.

DGHJGKJHGNSDFGH#################

[nix-shell:~/devel/BambooTracker/BambooTracker/BambooTracker]$ echo $?
1
```

tl;dr, BT *is* seemingly failing at gracefully handling an error. Maybe the error could be caught and handled better, would it be possible to continue without MIDI and open an error dialogue to inform the user? Like the one when the chosen sample rate is incompatible with the audio backend.

**Comment by jpcima** [comment]
>tl;dr, BT is seemingly failing at gracefully handling an error. Maybe the error could be caught and handled better, would it be possible to continue without MIDI and open an error dialogue to inform the user? Like the one when the chosen sample rate is incompatible with the audio backend.

This is the correct way to check for failure yes, use try/catch.
`error.what()` obtains the error message, if you would like to display it.

If it fails, you can then open RtMidi with the value `RtMidi::RTMIDI_DUMMY` for API.
This way it does not need to deal with null pointer checks.

**Comment by rerrahkr** [comment]
So should the tracker be able to change the api for MIDI In as well?
The tracker checks whether the specified api can be used with try/ catch, and sets `RtMidi::RTMIDI_DUMMY` when it cannot switch to the specified api.

**Comment by rerrahkr** [comment]
I added an error dialog at aac105643c88d49b6968db3c7d25ba071431d993.

**Comment by papiezak** [comment]
>I added an error dialog

I don't have MIDI keyboard, on Windows 10 I get this message every single time I open the tracker. The program works, but it got a bit annoying after while, is there a way to circumvent it? There is no need in removing that dialog.

**Comment by rerrahkr** [comment]
I changed initialization error check d16da7d866b5dc6f76ff3ee820d7dd9c5ec07e0b.

**Comment by titib** [comment]
Same issue.
I downloaded the app image then run it
I'm on Manjaro
5.15.167-1-MANJARO


```
ALSA lib seq_hw.c:531:(snd_seq_hw_open) open /dev/snd/seq failed: No such device

MidiOutAlsa::initialize: error creating ALSA sequencer client object.


MidiOutAlsa::initialize: error creating ALSA sequencer client object.

[1]    456634 segmentation fault (core dumped)  ./Downloads/PianoBooster-1.0.0-x86_64.AppImage

```

**Comment by OPNA2608** [comment]
@titib

> PianoBooster-1.0.0-x86_64.AppImage

Wrong issue tracker. This is for BambooTracker, and we don't offer AppImages.

Check if it was already reported at https://github.com/pianobooster/PianoBooster/issues or maybe already fixed, and open an issue there instead.
